### PR TITLE
Add google.protobuf.internal.containers.pyi to typeshed

### PR DESCRIPTION
**Pull request by nipunn1313**
We use this at Dropbox. We have prior permission from folks at google protobuf team to add stubs to mypy.

@dzbarsky did prior work and this is a port over from that.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Note that this fixes #1705 (right?) and the upstream protobuf project https://github.com/google/protobuf/issues/3803#issuecomment-341246195.

Note: we currently don't accept partial stubs for new packages. Are there other submodules in the google.protobuf package?

**Comment by nipunn1313** [comment]
This does part of #1705.
Is it possible to make an exception here for partial stubs, since there already exist partial stubs for protobuf?

```
nipunn-mbp:typeshed nipunn$ ls -R third_party/2/google/protobuf
__init__.pyi        descriptor.pyi      descriptor_pb2.pyi  descriptor_pool.pyi internal            message.pyi         message_factory.pyi reflection.pyi      symbol_database.pyi

third_party/2/google/protobuf/internal:
__init__.pyi          containers.pyi        decoder.pyi           encoder.pyi           enum_type_wrapper.pyi wire_format.pyi
```

And upstream in proto
```
nipunn-mbp:protobuf nipunn$ find python/google/protobuf/ -name "*.py"
python/google/protobuf//__init__.py
python/google/protobuf//descriptor.py
python/google/protobuf//descriptor_database.py
python/google/protobuf//descriptor_pool.py
python/google/protobuf//internal/__init__.py
python/google/protobuf//internal/_parameterized.py
python/google/protobuf//internal/api_implementation.py
python/google/protobuf//internal/containers.py
python/google/protobuf//internal/decoder.py
python/google/protobuf//internal/descriptor_database_test.py
python/google/protobuf//internal/descriptor_pool_test.py
python/google/protobuf//internal/descriptor_test.py
python/google/protobuf//internal/encoder.py
python/google/protobuf//internal/enum_type_wrapper.py
python/google/protobuf//internal/generator_test.py
python/google/protobuf//internal/import_test_package/__init__.py
python/google/protobuf//internal/json_format_test.py
python/google/protobuf//internal/message_factory_test.py
python/google/protobuf//internal/message_listener.py
python/google/protobuf//internal/message_test.py
python/google/protobuf//internal/proto_builder_test.py
python/google/protobuf//internal/python_message.py
python/google/protobuf//internal/reflection_test.py
python/google/protobuf//internal/service_reflection_test.py
python/google/protobuf//internal/symbol_database_test.py
python/google/protobuf//internal/test_util.py
python/google/protobuf//internal/testing_refleaks.py
python/google/protobuf//internal/text_encoding_test.py
python/google/protobuf//internal/text_format_test.py
python/google/protobuf//internal/type_checkers.py
python/google/protobuf//internal/unknown_fields_test.py
python/google/protobuf//internal/well_known_types.py
python/google/protobuf//internal/well_known_types_test.py
python/google/protobuf//internal/wire_format.py
python/google/protobuf//internal/wire_format_test.py
python/google/protobuf//json_format.py
python/google/protobuf//message.py
python/google/protobuf//message_factory.py
python/google/protobuf//proto_builder.py
python/google/protobuf//pyext/__init__.py
python/google/protobuf//pyext/cpp_message.py
python/google/protobuf//reflection.py
python/google/protobuf//service.py
python/google/protobuf//service_reflection.py
python/google/protobuf//symbol_database.py
python/google/protobuf//text_encoding.py
python/google/protobuf//text_format.py
```

We do seem to cover the important ones.

**Comment by gvanrossum** [comment]
Oh, I see -- there already were some stubs for protobuf, contributed by Google (337abed05a35f0c5b67c02d42bca0af1e2adc76e). I guess it's fine to add to that.

**Comment by nipunn1313** [comment]
@gvanrossum Sounds great! Anything further I should do to help with this PR?

File: third_party/2/google/protobuf/internal/containers.pyi [hunk 1]
```diff
@@ -0,0 +1,36 @@
+from google.protobuf.message import Message
+from typing import Sequence, TypeVar, Generic, Any, Iterator, Iterable, Union
+
+_T = TypeVar('_T')
+class BaseContainer(Generic[_T], Sequence[_T]):
```
Comment (JelleZijlstra) [inline third_party/2/google/protobuf/internal/containers.pyi]: You don't need to inherit specifically from Generic (but it's harmless, so feel free to leave this unchanged).

More importantly though, this class doesn't actually expose the entire Sequence interface; for example, there is no `.count()` method. Why don't you simply provide stubs for the methods that actually exist in the implementation?

File: third_party/2/google/protobuf/internal/containers.pyi [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    pass
+
+class RepeatedScalarFieldContainer(Generic[_T], BaseContainer[_T]):
```
Comment (JelleZijlstra) [inline third_party/2/google/protobuf/internal/containers.pyi]: Missing `__init__`.

File: third_party/2/google/protobuf/internal/containers.pyi [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    def append(self, value: _T) -> None: ...
+    def insert(self, key: int, value: _T) -> None: ...
+    def extend(self, elem_seq: Sequence[_T]) -> None: ...
+    def MergeFrom(self, other: RepeatedScalarFieldContainer[_T]) -> None: ...
+    def remove(self, elem: _T) -> None: ...
+    def pop(self, key: int = -1) -> _T: ...
```
Comment (JelleZijlstra) [inline third_party/2/google/protobuf/internal/containers.pyi]: Defaults should be `...`.

File: third_party/2/google/protobuf/internal/containers.pyi [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    def __iter__(self) -> Iterator[_T]: ...
+    def __delitem__(self, key: Union[int, slice]) -> None: ...
+    def __setitem__(self, key: int, value: _T) -> None: ...
+
+class RepeatedCompositeFieldContainer(Generic[_T], BaseContainer[_T]):
```
Comment (JelleZijlstra) [inline third_party/2/google/protobuf/internal/containers.pyi]: Also missing `__init__`.

**Comment by gvanrossum** [comment]
@nipunn1313 Can you respond to the code review?

**Comment by nipunn1313** [comment]
Apologies for the huge delay! Updated the diff here.

**Comment by nipunn1313** [comment]
I wasn't totally sure which ones could omit / keep the Generic[T]. It seems like we definitely need it for the base class, but not necessarily for the derived classes because they inherit from another class requiring a T, but I wasn't completely sure, so I left it in to be explicit.

**Comment by nipunn1313** [comment]
Hi @JelleZijlstra  - can you take another look?
Thanks

**Pull request merged**